src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+    /// </summary>
+    public SolutionCompilationState WithFrozenPartialCompilationIncludingSpecificDocument(
+        DocumentId documentId, CancellationToken cancellationToken)
+    {
+        try
+        {
+            var allDocumentIds = this.Solution.GetRelatedDocumentIds(documentId);
+            using var _ = ArrayBuilder<(DocumentState, SyntaxTree)>.GetInstance(allDocumentIds.Length, out var builder);
+
+            foreach (var currentDocumentId in allDocumentIds)
+            {
+                var document = this.Solution.GetRequiredDocumentState(currentDocumentId);
+                builder.Add((document, document.GetSyntaxTree(cancellationToken)));
+            }
+
+            using (this.StateLock.DisposableWait(cancellationToken))
+            {
+                // in progress solutions are disabled for some testing
+                if (Services.GetService<IWorkspacePartialSolutionsTestHook>()?.IsPartialSolutionDisabled == true)
+                {
+                    return this;
+                }
+
+                SolutionCompilationState? currentPartialSolution = null;
+                _latestSolutionWithPartialCompilation?.TryGetTarget(out currentPartialSolution);
+
+                var reuseExistingPartialSolution =
+                    (DateTime.UtcNow - _timeOfLatestSolutionWithPartialCompilation).TotalSeconds < 0.1 &&
+                    _documentIdOfLatestSolutionWithPartialCompilation == documentId;
+
+                if (reuseExistingPartialSolution && currentPartialSolution != null)
+                {
+                    SolutionLogger.UseExistingPartialSolution();
+                    return currentPartialSolution;
+                }
+
+                var newIdToProjectStateMap = this.Solution.ProjectStates;
+                var newIdToTrackerMap = _projectIdToTrackerMap;
+
+                foreach (var (doc, tree) in builder)
+                {
+                    // if we don't have one or it is stale, create a new partial solution
+                    var tracker = this.GetCompilationTracker(doc.Id.ProjectId);
+                    var newTracker = tracker.FreezePartialStateWithTree(this, doc, tree, cancellationToken);
+
+                    Contract.ThrowIfFalse(newIdToProjectStateMap.ContainsKey(doc.Id.ProjectId));
+                    newIdToProjectStateMap = newIdToProjectStateMap.SetItem(doc.Id.ProjectId, newTracker.ProjectState);
+                    newIdToTrackerMap = newIdToTrackerMap.SetItem(doc.Id.ProjectId, newTracker);
+                }
+
+                var newState = this.Solution.Branch(
+                    idToProjectStateMap: newIdToProjectStateMap,
+                    dependencyGraph: SolutionState.CreateDependencyGraph(this.Solution.ProjectIds, newIdToProjectStateMap));
+                var newCompilationState = this.Branch(
+                    newState,
+                    newIdToTrackerMap);
```

Comment:
a case wheer we first branch the solution state, then branch the compilation state.